ocaml/typing/typedecl.ml

```diff
@@ -1155,6 +1207,7 @@ let check_kind_coherence env loc dpath decl =
   | _ -> ()
 
 let check_coherence env loc dpath decl =
+  let decl = reexport_special_types env loc dpath decl in
```

Comment:
Why do this in `check_coherence`? I was expecting this to appear in `transl_declaration` directly, but maybe there's a good reason for it to be here.

Reply:
We must re-export `or_null` after `check_constraints`, because `check_constraints` expects `type_kind` to match the one in the parsetree.

Reply:
Firstly, that's terrible. `check_constraints` is pretty sad. But I suppose we don't need to fix it today.

But we can just skip this. An `or_null` re-export doesn't need to check constraints: I don't think there's any way for it to fail. `check_constraints` is needed when you have mutually recursive type declarations where one type declares constraints that the other types might or might not respect. That's just not the case for `or_null` re-exports: we know, for sure, that neither the manifest nor the kind mentions any of the types brought into scope within this same group. So it's all OK.

My suggestion: check for the `[@@or_null_reexport]` attribute in `check_constraints` and just do nothing if that attribute is present.

Reply:
Done. Regarding shapes, I ended up modifying `shape_map_cstrs` to accept `Types.constructor_declaration`, since it has all necessary information.